File: backend/src/main/java/com/icr/backend/service/UserService.java
package com.icr.backend.service;

import com.icr.backend.dto.request.CreateUserRequest;
import com.icr.backend.dto.request.ResetPasswordRequest;
import com.icr.backend.dto.response.FacultyStudentAnalyticsResponse;
import com.icr.backend.dto.response.PageResponse;
import com.icr.backend.dto.response.UserPublicDTO;
import com.icr.backend.dto.response.UserResponse;
import com.icr.backend.dto.response.DashboardStatsResponse;
import com.icr.backend.dto.response.UserStatsResponse;

import java.util.List;


public interface UserService {

    PageResponse<UserResponse> getAllUsers(int page, int size, String role);
    UserStatsResponse getUserStats();

    UserResponse getUserById(Long id);

    void deleteUser(Long id);

    DashboardStatsResponse getDashboardStats();

    UserResponse createUser(CreateUserRequest request);

    void resetPassword(Long userId, ResetPasswordRequest request);

    void reassignStudent(Long studentId, Long newFacultyId);
    PageResponse<UserResponse> getStudentsByFaculty(Long facultyId, String status, int page, int size);
    List<FacultyStudentAnalyticsResponse> getFacultyStudentAnalytics();
    List<UserPublicDTO> getApprovedFacultyPublic();

    List<UserResponse> getAllFaculty();

    List<UserResponse> getPendingFaculty();

    UserResponse approveFaculty(Long facultyId);

    UserResponse rejectFaculty(Long facultyId, String reason);

    List<UserResponse> getPendingStudents();
    List<UserResponse> getAssignedStudents();

    UserResponse approveStudent(Long studentId);

    UserResponse rejectStudent(Long studentId, String reason);

    UserResponse getRegistrationStatus(String email);

}

File: frontend/src/api/userService.js
import axiosInstance from "./axiosInstance";

export const getUsers = async ({ page = 0, size = 5, role = null }) => {
  const params = { page, size };
  if (role) params.role = role;
  const response = await axiosInstance.get("/users", { params });
  return response.data?.data;
};

export const getUserStats = async () => {
  const response = await axiosInstance.get("/users/stats");
  return response.data?.data ?? { total: 0, faculty: 0, student: 0, admin: 0 };
};

export const deleteUser = async (id) => {
  await axiosInstance.delete(`/users/${id}`);
};

export const createUser = async (userData) => {
  const response = await axiosInstance.post("/admin/users", userData);
  return response.data;
};

export const resetUserPassword = async ({ userId, newPassword }) => {
  const response = await axiosInstance.put(`/admin/users/${userId}/reset-password`, {
    newPassword,
  });
  return response.data;
};

export const getFacultyList = async () => {
  const response = await axiosInstance.get("/users/faculty");
  return response.data?.data ?? response.data ?? [];
};

export const reassignStudent = async ({ studentId, newFacultyId }) => {
  const response = await axiosInstance.put(
    `/admin/students/${studentId}/reassign`,
    null,
    { params: { newFacultyId } }
  );
  return response.data;
};

// Admin — faculty approvals
export const getPendingFaculty = async () => {
  const response = await axiosInstance.get("/admin/pending-faculty");
  return response.data;
};

export const approveFaculty = async (id) => {
  const response = await axiosInstance.put(`/admin/faculty/${id}/approve`);
  return response.data;
};

export const rejectFaculty = async ({ id, reason }) => {
  const response = await axiosInstance.put(`/admin/faculty/${id}/reject`, null, {
    params: { reason },
  });
  return response.data;
};

// Faculty — student approvals
export const getPendingStudents = async () => {
  const response = await axiosInstance.get("/faculty/pending-students");
  const payload = response.data;
  if (Array.isArray(payload)) return payload;
  if (Array.isArray(payload?.data)) return payload.data;
  if (Array.isArray(payload?.content)) return payload.content;
  if (Array.isArray(payload?.data?.content)) return payload.data.content;
  return [];
};

export const approveStudent = async (id) => {
  const response = await axiosInstance.put(`/faculty/students/${id}/approve`);
  return response.data;
};

export const rejectStudent = async ({ id, reason }) => {
  const response = await axiosInstance.put(`/faculty/students/${id}/reject`, null, {
    params: { reason },
  });
  return response.data;
};

export const getFacultyStudentAnalytics = async () => {
  const response = await axiosInstance.get("/admin/students/faculty-analytics");
  return Array.isArray(response.data) ? response.data : [];
};

export const getStudentsByFaculty = async ({
  facultyId,
  page = 0,
  size = 10,
  status = "ALL",
}) => {
  const response = await axiosInstance.get(`/admin/students/faculty/${facultyId}`, {
    params: { page, size, status },
  });
  return response.data;
};
